### PR TITLE
Update zlib tests to show shared and static and external rebuild behaviors.

### PR DESCRIPTION
**Pull request by kquick**

### CODE REVIEW

File: tests/common/mod.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+//
+// If the user did not provide one, return None, which build-bom interprets as 'llvm-link'
+pub fn user_llvm_link_cmd() -> Option<String> {
+    std::env::var("LLVM_LINK").ok()
+}
+
+// Get the user-provided llvm-dis command (via the LLVM_DIS environment variable), if any
+//
+// If the user did not provide one, return None, which build-bom interprets as 'llvm-dis'
+#[allow(dead_code)]
```
Comment (travitch) [inline tests/common/mod.rs:25]: Is this directive necessary? It looks like it is used later

File: tests/common/mod.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+use bom;
+use bom::bom::options::{BitcodeOptions, ExtractOptions, Options, Subcommand};
+use std::path::{PathBuf};
+
+pub static SOURCE_DIR: &'static str = "tests/sources";
+
+
+// Get the clang command provided by the user (via the CLANG environment variable), if any
+//
+// If the user did not provide one, just return None, which is interpreted by build-bom as 'clang'
```
Comment (travitch) [inline tests/common/mod.rs:10]: If these commands are used just in the tests, does it make sense to have them return `PathBuf`/`String` and just do the default replacement here?

File: tests/test_zlib.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        let cmd = Cmd::new("wget").arg("-O").arg(path.as_path()).arg(url);
+        cmd.run()?;
+    }
+
+    let abs_path = std::fs::canonicalize(path.as_path())?;
+    Ok(abs_path)
+}
+
+
+#[allow(dead_code)]  // tdir is holding a resource, but never directly used
```
Comment (travitch) [inline tests/test_zlib.rs:39]: Can you just prefix its name with an underscore to get rid of that warning?

File: tests/test_zlib.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                panic!("Unable to compile zlib for testing: {:?}", e),
+        }
+    };
+}
+
+
+fn zlib_do_build() -> anyhow::Result<(TempDir, String, PathBuf)> {
+    let url = "https://www.zlib.net/current/zlib.tar.gz";
+    let filename = "zlib.tar.gz";
+    let abs_src = fetch_if_needed(url, filename)?;
```
Comment (travitch) [inline tests/test_zlib.rs:61]: This can be a different PR, but I do not have any idea why I did it this way... it seems like I should have just checked in a copy with git-lfs?

File: tests/test_zlib.rs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        cpopts.content_only = true;
+        copy(tgt_path, my_tgt_path, &cpopts)?;
+
+        // While the above copy is intended to preserve the file timestamps, it
+        // is not always guaranteed that there will be enough system time
+        // resolution to preserve the "make finished" status, so re-run the build
+        // via build-bom to ensure it is up-to-date before making any changes.
+        let cmd_opts = vec![String::from("make"),
+                            // "-C".to_string(),
+                            // String::from(my_tgt_path.to_str().unwrap())
```
Comment (travitch) [inline tests/test_zlib.rs:271]: Delete? I assume this is no longer necessary because of `serial`?

File: tests/test_zlib.rs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+                                        command: cmd_opts,
+                                        any_fail: false };
+        {
+            let _push2 = pushd(my_tgt_path)?;
+            common::gen_bitcode(gen_opts.clone())?;
+        }
+
+        let mut exe_path = PathBuf::from(my_tgt_path);
+        exe_path.push("example64");
+        assert!(exe_path.exists());  // build successfully generated a static executable
```
Comment (travitch) [inline tests/test_zlib.rs:289]: You can add an extra argument to the `assert!` macro that is a string describing the assertion (instead of the comment).

File: tests/test_zlib.rs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+            ll.contains("inflateCopy"),   // static lib
+            ll.contains("compressBound"),   // static lib
+            ll.contains("deflateSetHeader"),   // static lib
+            ll.contains("gzfread"),   // static lib, rebuilt outside of build-bom
+            ll.contains("gzfwrite"),   // static lib
+            ll.contains("test_gzio"),  // exe src
+        ];
+        assert_eq!(fnd, vec![true, true, true, false, true, true]);
+    }
+    assert!(tdir.path().exists()); // ensure tempdir lasts until here
```
Comment (travitch) [inline tests/test_zlib.rs:329]: Is this testing that build-bom didn't delete it? If you are just ensuring that the test harness isn't broken, you could use the `mem::drop` trick here too

File: tests/common/mod.rs (hunk 1 above)
Comment (kquick) [reply tests/common/mod.rs:25]: Each test file becomes a separate crate, so while this is used in test_zlib.rs, it isn't used in test_bom.rs, so it generates a warning during the independent compile of the latter without this directive.

File: tests/common/mod.rs (hunk 2 above)
Comment (kquick) [reply tests/common/mod.rs:10]: Makes sense for the `user_llvm_dis_cmd`, and I changed that, but the others are actual `Option` fields in the options structures.

File: tests/test_zlib.rs (hunk 3 above)
Comment (kquick) [reply tests/test_zlib.rs]: This was actually obsolete because I do have an explicit reference in each of the tests, so I removed it.  The main concern is that it represents the tmpdir which will be removed when this is dropped, so it shouldn't be dropped early. I would have been worried that an underscore prefix and lack of reference would cause rust to drop it if optimized, but per the above, this isn't needed anymore.

File: tests/test_zlib.rs (hunk 4 above)
Comment (kquick) [reply tests/test_zlib.rs:60]: The file *is* actually checked in (https://github.com/travitch/build-bom/blob/master/tests/sources/zlib.tar.gz), but something frequently deletes it and it has to be refreshed.  I haven't tracked down what is deleting it, but this is not usually causing much of a problem.

File: tests/test_zlib.rs (hunk 5 above)
Comment (kquick) [reply tests/test_zlib.rs]: Doh! Yes, `serial` +  `pushd` obsoleted this.

File: tests/test_zlib.rs (hunk 6 above)
Comment (kquick) [reply tests/test_zlib.rs]: Good idea... done.

File: tests/test_zlib.rs (hunk 7 above)
Comment (kquick) [reply tests/test_zlib.rs]: Actually, this is a static that is shared by all 5 tests, so a particular test cannot/should not force it to be dropped because it will still be used by the others.  The reference at the end is just a bit of additional insurance that no optimization or accidental move caused it to be dropped earlier.

File: tests/common/mod.rs (hunk 1 above)
Comment (kquick) [reply tests/common/mod.rs:25]: I suppose I could move it into test_zlib.rs, but that would move it away from the other two similar lookups.  Let me know which you prefer.

Comment (travitch) [reply tests/common/mod.rs:25]: Weird. It would be worth a note there at least. No need for big changes.
